### PR TITLE
feat(transaction): generate past occurrences for recurring transactions

### PR DESCRIPTION
**Pull request by Ic3b3rg**
## 🎯 Description

This PR adds functionality to automatically generate all past transactions when a recurring transaction is created with a start date in the past.
#Problem
Currently, when users create a recurring transaction with a start date in the past, the system doesn't automatically generate the transactions that should have occurred between the start date and the current date. This leads to incomplete transaction history and inaccurate financial records.

#Solution
Added logic to detect when a recurring transaction's start date is in the past
Implemented calculation of all dates between the start date and current date based on the recurrence pattern (daily, weekly, monthly, etc.)
Automatically creates individual transactions for each calculated past date
Maintains proper transaction records with correct recurring transaction relationships
This enhancement improves data integrity and saves users from having to manually create past recurring transactions.

Closes: #336   

## 📱 Changes

- [x] Describe key changes made
- [x] List any new features, bug fixes, or refactors
- [x] Include additional details if necessary

### Behaviour
Behaviour
1. Create a new recurring transaction with a start date in the past (e.g., 3 months ago)
2. Select different recurrence types (daily, weekly, monthly) to verify all work correctly
3. After saving, navigate to the transactions list and verify:
- All expected past transactions have been created
- The dates follow the correct recurrence pattern
- Each transaction is properly linked to the recurring transaction
4. Verify the transactions appear correctly in reports and balance calculations
5. Try an edge case with a long past date (e.g., 1 year ago) with different recurrence patterns
#Edge Cases
1. Test with months having different number of days (e.g., start date on 31st of a month)
2. Test with dates spanning across leap years
3. Verify behavior when creating a recurring transaction with both start and end dates in the past

## 📸 Screenshots / Screen Recordings (if applicable)

If your PR involves UI changes, please add screenshots or screen recordings here.

| Platform | Before | After |
|----------|--------|-------|
| Android  |        |       |
| iOS      |        |       |

## 🔍 Checklist for reviewers
- [ ] Code is formatted correctly
- [ ] Tests are passing
- [ ] New tests are added (if needed)
- [ ] Style matches the figma/designer requests
- Tested on:
    - [ ] iOS
    - [ ] Android

## ✍️ Additional Context

Add any other information that we might know

### CODE REVIEW

File: lib/providers/transactions_provider.dart [hunk 1]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+    }
+  }
+
+  // Verifica se la data è oggi
+  bool _isToday(DateTime date) {
+    DateTime now = DateTime.now();
+    return date.year == now.year &&
+        date.month == now.month &&
+        date.day == now.day;
+  }
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:278]: There's already an extension on `DateTime` that does this:

https://github.com/RIP-Comm/sossoldi/blob/a931c3966d06bd802dd494f44c28735044396296/lib/ui/extensions.dart#L73-L75

Can you please use this?

File: lib/providers/transactions_provider.dart [hunk 2]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+    if (originalDay > lastDayOfNextPeriod) {
+      dayOfInsertion = lastDayOfNextPeriod;
+    }
+
+    // Crea la nuova data
+    return DateTime(
+        currentDate.year, currentDate.month + monthsToAdd, dayOfInsertion);
+  }
+
+  // Crea una singola transazione
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:322]: This comment is avoidable.

File: lib/providers/transactions_provider.dart [hunk 3]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+        idBankAccount: idBankAccount,
+        idCategory: idCategory,
+        recurrency: recurrency,
+        type: type.name,
+        createdAt: date,
+        updatedAt: date,
+        lastInsertion: date);
+  }
+
+  // Genera le transazioni individuali per una transazione ricorrente
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:216]: This comment is avoidable too

File: lib/providers/transactions_provider.dart [hunk 4]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+      DateTime startDate,
+      Recurrence recurrency) async {
     DateTime now = DateTime.now();
 
-    if (date.year == now.year &&
-        date.month == now.month &&
-        date.day == now.day) {
-      final transaction = Transaction(
-        date: date,
+    // Se la data di inizio è nel passato, genera tutte le transazioni passate
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:227]: Same here

File: lib/providers/transactions_provider.dart [hunk 5]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+    if (startDate.isBefore(now)) {
+      await _createPastTransactions(startDate, now, recurrency,
+          recurringTransaction.id!, amount, label, idBankAccount, idCategory);
+    }
+    // Se la data è oggi, crea solo la transazione corrente
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:232]: Same here

File: lib/providers/transactions_provider.dart [hunk 6]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+      } else if (entity == 'months') {
+        currentDate =
+            _calculateNextMonthDate(currentDate, intervalAmount, startDate.day);
+      }
+    }
+
+    return dates;
+  }
+
+  // Calcola la prossima data mensile considerando il numero di giorni nel mese
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:303]: Can you please translate all the relevant comment in English?

File: lib/providers/transactions_provider.dart [hunk 7]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+        DateTime(currentDate.year, (currentDate.month + monthsToAdd + 1), 0)
+            .day;
+
+    // Se il giorno originale è maggiore del numero di giorni nel mese, usa l'ultimo giorno
+    int dayOfInsertion = originalDay;
+    if (originalDay > lastDayOfNextPeriod) {
+      dayOfInsertion = lastDayOfNextPeriod;
+    }
+
+    // Crea la nuova data
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:317]: Avoidable comment also here

File: lib/providers/transactions_provider.dart [hunk 8]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+
+  // Calcola la prossima data mensile considerando il numero di giorni nel mese
+  DateTime _calculateNextMonthDate(
+      DateTime currentDate, int monthsToAdd, int originalDay) {
+    // Calcola il giorno valido per il mese successivo
+    int lastDayOfNextPeriod =
+        DateTime(currentDate.year, (currentDate.month + monthsToAdd + 1), 0)
+            .day;
+
+    // Se il giorno originale è maggiore del numero di giorni nel mese, usa l'ultimo giorno
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:311]: Tranlsate also here

File: lib/providers/transactions_provider.dart [hunk 9]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+      }
+    }
+
+    return dates;
+  }
+
+  // Calcola la prossima data mensile considerando il numero di giorni nel mese
+  DateTime _calculateNextMonthDate(
+      DateTime currentDate, int monthsToAdd, int originalDay) {
+    // Calcola il giorno valido per il mese successivo
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:306]: Tranlsate also here

File: lib/providers/transactions_provider.dart [hunk 10]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+  Future<void> _createPastTransactions(
+      DateTime startDate,
+      DateTime endDate,
+      Recurrence recurrency,
+      int recurringTransactionId,
+      num amount,
+      String label,
+      int idBankAccount,
+      int idCategory) async {
+    // Calcola tutte le date passate in cui creare transazioni
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:255]: Avoidable comment here

File: lib/providers/transactions_provider.dart [hunk 11]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+    List<DateTime> transactionDates =
+        _calculatePastTransactionDates(startDate, endDate, recurrency);
+
+    // Crea una transazione per ogni data calcolata
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:259]: Tranlsate here

File: lib/providers/transactions_provider.dart [hunk 12]
```diff
@@ -159,49 +159,187 @@ class AsyncTransactionsNotifier
...
+        idBankAccount: idBankAccount,
+        idCategory: idCategory,
+        idRecurringTransaction: recurringTransaction.id!,
       );
-      await TransactionMethods().insert(transaction);
     }
+  }
 
-    return insertedTransaction;
+  // Crea tutte le transazioni per date passate
```
Comment (fres-sudo) [inline lib/providers/transactions_provider.dart:245]: Avoidable comment here

**Review by fres-sudo — CHANGES_REQUESTED** [review]
Didn't test the code yet, just added some reviews.

Since this is a sensitive function that is trivial to debug if there are some errors, can you please add tests for it?

**Comment by lucaantonelli** [comment]
I think the PR miss some files, the type argument is added in recurringtransaction but is not in the recurring transaction model, the app doesn't run with this pr actually.

**Comment by Ic3b3rg** [comment]
@fres-sudo i will delete all comments, i think that the name of functions already clear

**Comment by theperu** [comment]
Hi @Ic3b3rg👋
We recently merged a PR that changed the folder structure, which caused some conflicts here. When you get a chance, could you take a look and resolve them?

If there’s no response in the next 15 days, we’ll consider this PR inactive and plan to close it to keep things tidy.

Thanks!

**Comment by Ic3b3rg** [comment]
Hi, due to a lot of commitments I haven’t been able to make progress on this. I tried to look into the conflicts, but something completely broke the branch even though I managed to resolve them.
I’d suggest closing this PR and reopening the issue on GitHub as unassigned, since at the moment I’m not able to work on it.
